[Mobile] Use link picker in Image block

mkevins opened the pull request:
#### Related PRs

`gutenberg-mobile`: https://github.com/wordpress-mobile/gutenberg-mobile/pull/2841
`WordPress-Android`: https://github.com/wordpress-mobile/WordPress-Android/pull/13576

## Description

This PR adds the link picker functionality to the Image block settings, and auto-hides the "Open in new tab" and "rel" settings when a link is not present. This is a first step towards resolving: https://github.com/wordpress-mobile/gutenberg-mobile/issues/1655.

This also auto-hides the "Open in new tab" setting for RichText based link editing.

## Testing steps

#### Link picker

* Add an image to a post
* Tap the block settings gear
* Tap "Link To: Search or type URL" cell
  * Expect link picker to open
* Type some text and pick a link from the results
  * Expect the block settings screen to return with the selected link URL set
  * Expect "Open in new tab" and "rel" settings to appear
* Tap the link URL to re-open the link picker
* Delete the link and tap the :white_check_mark: check mark to confirm
  * Expect the "Open in new tab" and "rel" settings to be hidden in the block settings

#### Open in new tab

* Tap the settings gear on an image with a link set (from previous steps)
* Toggle the "Open in new tab" switch
  * Expect the setting to toggle
  * Expect the "Link Rel" setting to have "noreferrer noopener" toggled as well

#### RichText auto-hiding

* Add a paragraph block and type some text
* Highlight some text and tap the link button
  * Expect that "Open in new tab" option is not present
* Add a link from the link picker
  * Expect that "Open in new tab" (and Remove link) options are present

## Screenshots <!-- if applicable -->

<details>
<summary>Image block link settings</summary>

Link picker | Open in new tab
-|-
![image-block-add-link](https://user-images.githubusercontent.com/8507675/100419090-8e60bc80-30cf-11eb-8159-401031437eda.gif)|![image-block-toggle-open-in-new-tab](https://user-images.githubusercontent.com/8507675/100419097-902a8000-30cf-11eb-90f8-23fddb1e16c0.gif)

Note: the above screen recordings were created prior to autohiding the "Open in new tab" and "rel" settings.
</details>

<details>
<summary>Autohide settings</summary>

Image block | Button block
-|-
![image-block-settings-autohide](https://user-images.githubusercontent.com/8507675/101751803-700dae80-3b1c-11eb-8977-aba48452f250.gif)|![button-block-settings-autohide](https://user-images.githubusercontent.com/8507675/101751797-6d12be00-3b1c-11eb-900e-3148c50394d8.gif)

</details>

## Types of changes
New feature (non-breaking change which adds functionality)

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/accessibility/ -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://developer.wordpress.org/coding-standards/inline-documentation-standards/javascript/ -->
- [x] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR. <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/master/docs/contributors/native-mobile.md -->

mkevins committed:
Use LinkSettingsNavigation in Image block

mkevins committed:
Revert mapping url attribute in link settings

mkevins committed:
Map url -> href attribute for link picker in image block

mkevins committed:
Don't overwrite mapped attribute when undefined

mkevins committed:
Fix lint

mkevins commented on packages/block-library/src/image/edit.native.js:
```diff
@@ -338,6 +336,49 @@ export class ImageEdit extends React.Component {
+			setAttributes,
+		} = this.props;
+
+		const mappedAttributes = { ...unMappedAttributes, url };
+		const setMappedAttributes = ( { url: href, ...restAttributes } ) =>
+			href === undefined
+				? setAttributes( restAttributes )
+				: setAttributes( { ...restAttributes, href } );
```
The link settings sets the attribute named `url`, but Image block's `url` attribute is the URL of the image itself, and instead uses the attribute named `href` to represent the link URL.

Originally, to work around this, I used [dynamic properties](https://github.com/WordPress/gutenberg/pull/27292/commits/c7cf5873823a880866facc03e290e6fce9c3c7ad#diff-1c6b716518b93eef13a216ed21fa8d92122f8941fbac8378f13f73a4ae5314dc) in `LinkSettings` to allow the consumer to customize the attribute name to be set.

Here, I reverted those changes to `LinkSettings`, and instead map the attibutes in `ImageEdit`. As the link picker is used in more places, it might make sense to re-evaluate this pattern, depending on the commonality of the possible attribute names (e.g. `url`, `href`). WDYT?

enejb replied:
I like the dynamic properties approach. It seems fairly clean. The code ^ is a bit harder for me to understand. But I think this could be done as part of a larger refactor.

One thing that I noticed is that we always seem to set the options such as labels etc. 

I think we could really benefit from having a default name for labels. so that we only pass in the options that are different from the default.

mkevins replied:
> I think we could really benefit from having a default name for labels. so that we only pass in the options that are different from the default.

Exactly my thoughts! I think we can refactor during the iteration for further consolidation [described here](https://github.com/wordpress-mobile/gutenberg-mobile/issues/1655#issuecomment-745619398). I think with the refactor we should have the opportunity to both reduce the surface area of the interface, and also make the components appropriately general (in particular, decoupled from particular attribute names and even `setAttributes`).

mkevins committed:
Merge branch 'master' into rnmobile/try/image-block-link-picker

mkevins committed:
Conditionally render link options in link settings

mkevins committed:
Move link settings panel in image block settings

mkevins committed:
Auto-hide "Open in new tab" in rich text link settings

mkevins committed:
Merge branch 'master' into rnmobile/try/image-block-link-picker

enejb commented:
This PR worked really well for me. 

I tested it on the  iOS simulator. I think the LinkSettings could use some better default options so that it would be harder to set the link setting labels to be different across the different instances.

enejb reviewed (approved):
Lets ship this and refactor how the options that we pass into the `LinkSettingsNavigation` to have better defaults.

mkevins committed:
Merge branch 'master' into rnmobile/try/image-block-link-picker

mkevins committed:
Merge branch 'master' into rnmobile/try/image-block-link-picker

mkevins committed:
Revert "Update tt1 blocks references. (#27770)"

This reverts commit c665998a2ba5cdb9d086cb91aad8c621f44af49d.

mkevins commented:
Sorry for the ping.. I pushed a revert to troubleshoot the failing test: https://github.com/WordPress/gutenberg/blame/master/packages/e2e-tests/specs/experiments/multi-entity-saving.test.js#L86.

mkevins committed:
Revert "Revert "Update tt1 blocks references. (#27770)""

This reverts commit f40b256f46bccdf672b28e56dc63e7bf5123dda7.

mkevins committed:
Fix assertion count in multi-entity-saving test

This is an attempt to fix a failing test - I manually counted 26
assertions, instead of 27, so it looks correct. However, it still is not
clear how this was passing before with 27.

mkevins committed:
Merge branch 'master' into rnmobile/try/image-block-link-picker

mkevins commented:
Thank you Enej for reviewing and testing! :smiley:
